File: DFS/695.max-area-of-island.go
package leetcode

// @leet start
type p2 struct {
	r, c int
}

func (p p2) move(p1 p2) p2 {
	return p2{p.r + p1.r, p.c + p1.c}
}

func maxAreaOfIsland(grid [][]int) int {
	var (
		nr, nc = len(grid), len(grid[0])
		dirs   = [4]p2{
			{0, 1},
			{0, -1},
			{-1, 0},
			{1, 0},
		}
		res int
	)
	isInside := func(p p2) bool {
		return p.r >= 0 && p.r < nr && p.c >= 0 && p.c < nc
	}

	var dfs func(p p2) int
	dfs = func(p p2) int {
		if !isInside(p) || grid[p.r][p.c] != 1 {
			return 0
		}
		grid[p.r][p.c] = 0
		res := 0
		for _, dir := range dirs {
			res += dfs(p.move(dir))
		}
		return res + 1
	}

	for r := range nr {
		for c := range nc {
			p := p2{r, c}
			if grid[r][c] == 0 {
				continue
			}
			res = max(res, dfs(p))
		}
	}
	return res
}

// @leet end

File: DFS/695.max-area-of-island.py
from typing import List
from typing import NamedTuple


# @leet start
class Pt(NamedTuple):
    r: int
    c: int

    def mv(self, d: "Pt") -> "Pt":
        return Pt(self.r + d.r, self.c + d.c)


class Solution:
    dirs = [
        Pt(1, 0),
        Pt(-1, 0),
        Pt(0, -1),
        Pt(0, 1),
    ]

    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        ret = 0
        nr, nc = len(grid), len(grid[0])

        def inside(p: Pt):
            return 0 <= p.r < nr and 0 <= p.c < nc

        def dfs(p: Pt):
            if not inside(p) or grid[p.r][p.c] != 1:
                return 0
            ret = 1
            grid[p.r][p.c] = 2
            for d in self.dirs:
                ret += dfs(p.mv(d))
            return ret

        for r in range(nr):
            for c in range(nc):
                if grid[r][c] == 1:
                    p = Pt(r, c)
                    ret = max(ret, dfs(p))

        return ret


# @leet end

